runtime/service.go

```diff
@@ -489,6 +490,54 @@ func dialVsock(ctx context.Context, contextID uint32, port uint32) (net.Conn, er
 	return nil, lastErr
 }
 
+// findNextAvailableVsockCID finds first available vsock context ID.
+// It uses VHOST_VSOCK_SET_GUEST_CID ioctl which allows some CID ranges to be statically reserved in advance.
+// The ioctl fails with EADDRINUSE if cid is already taken and with EINVAL if the CID is invalid.
+// Taken from https://bugzilla.redhat.com/show_bug.cgi?id=1291851
+func findNextAvailableVsockCID(ctx context.Context) (uint32, error) {
+	const (
+		// Corresponds to VHOST_VSOCK_SET_GUEST_CID in vhost.h
+		ioctlVsockSetGuestCID = uintptr(0x4008AF60)
+		// 0, 1 and 2 are reserved CIDs, see http://man7.org/linux/man-pages/man7/vsock.7.html
+		startCID = 3
+		maxCID   = math.MaxUint32
+	)
+
+	file, err := os.OpenFile("/dev/vhost-vsock", syscall.O_RDWR, 0666)
+	if err != nil {
+		return 0, errors.Wrap(err, "failed to open vsock device")
+	}
+
+	defer file.Close()
+
+	for contextID := startCID; contextID < maxCID; contextID++ {
+		select {
+		case <-ctx.Done():
+			return 0, ctx.Err()
+		default:
+			cid := contextID
+			_, _, err = unix.Syscall(
```

Comment:
If we allow mocking of the `Syscall`, we can test this. How do you feel about adding a small abstraction that will allow for you to call like `var unixSyscall = unix.Syscall` then in the tests, set the `unixSyscall` to a mock function.

```go
func TestMock(t *testing.T) {
    unixSyscall = func(/*parameters*/) (uintptr, uintptr, syscall.Errno) {
        return 0, 0, syscall.Errno(0)
    }
}
}
```